### PR TITLE
changefeedccl: support primary key changes

### PR DESCRIPTION
**Pull request by ajwerner**
This PR does a few things. One is that when errors occur due to unsupported
schema changes during the execution of a changefeed, relatively poor handling
ensues. Ideally we'd allow the changefeed to run its course right up to that
unsupported schema change's timestamp and then ensure that we persist the fact
that we've processed all of that data. That would permit a user to then restart
a changefeed after the unsupported change. There are some edge cases here worth
considering related to off-by-ones in the timestamp management. I probably
should go through that exercise before merging this PR.

The real feature this work is in support of is to allow for changefeeds to
successfully navigate changes to a primary index.

This PR works and support changes to the primary key of a table that also
include column set changes.

Release note (enterprise change): Support primary key changes in `CHANGEFEED`.

### CODE REVIEW

File: pkg/jobs/jobspb/jobs.proto [hunk 1]
```diff
@@ -462,7 +462,34 @@ message ChangefeedDetails {
...
+
+  // BoundaryReached indicates that this resolved timestamp
+  // corresponds to a schema change boundary. A schema change boundary is a
+  // timestamp at which a new version of a table has been committed that either
+  // indicates a logical backfill or a need for the changefeed to exit.
+  bool deprecated_boundary_reached = 3;
+
+  enum BoundaryType {
+
+    // NONE indicates that this this resolved span does not correspond to a
```
Comment (HonoreDB) [inline pkg/jobs/jobspb/jobs.proto:474]: "this this"

File: pkg/jobs/jobspb/jobs.proto [hunk 2]
```diff
@@ -462,7 +462,34 @@ message ChangefeedDetails {
...
+    // requiring a backfill internally and perhaps indicates the need for a
+    // protected timestamp.
+    BACKFILL = 1;
+
+    // EXIT indicates that this resolved span corresponds to a boundary which
+    // should result in the changefeed exiting.
+    EXIT = 2;
+
+    // RESTART indicates that this resolved span corresponds to a boundary which
+    // should result in the changefeed exiting.
```
Comment (HonoreDB) [inline pkg/jobs/jobspb/jobs.proto:488]: "should result in the changefeed restarting"

File: pkg/ccl/changefeedccl/changefeed_processors.go [hunk 3]
```diff
@@ -710,6 +735,15 @@ func (cf *changeFrontier) noteResolvedSpan(d rowenc.EncDatum) error {
 			`unmarshalling resolved span: %x`, raw)
 	}
 
+	// TODO(ajwerner): Remove this code in 21.2.
+	if resolved.DeprecatedBoundaryReached && resolved.BoundaryType == jobspb.ResolvedSpan_NONE {
```
Comment (HonoreDB) [inline pkg/ccl/changefeedccl/changefeed_processors.go:739]: I'm confused about how this conditional is ever true. Why is `DeprecatedBoundaryReached` already true if `resolved.BoundaryType` is none?

**Review by HonoreDB — COMMENTED** [review]
Thanks for pushing this out! It has implications, I think, for the simpler stuff I've been trying to get done here like tolerating table name changes. Just adding some comment nits for now.

File: pkg/ccl/changefeedccl/changefeed_processors.go (hunk 3 above)
Comment (ajwerner) [reply pkg/ccl/changefeedccl/changefeed_processors.go:739]: Needs commentary. This code also is not tested :grimacing:. The idea is that nodes running the old code will only have set the boolean and not the type but the below code only wants to deal in types. I'll think about how I want to test this.

**Comment by ajwerner** [comment]
@HonoreDB any chance I could get a review on this one to get it into this release? @miretskiy adding you too

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 739 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MQDnWAKTpM3zyX68mnf-r1-739:-MTu_x_g7BBKbJfd9MRA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebf2c85295ce98718e7df61fc28589f0aeaa73e/pkg/ccl/changefeedccl/changefeed_processors.go#L739)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Needs commentary. This code also is not tested :grimacing:. The idea is that nodes running the old code will only have set the boolean and not the type but the below code only wants to deal in types. I'll think about how I want to test this. 
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/jobs.proto, line 474 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MQDnWAKTpM3zyX68mnp-r1-474:-MTu_yZ2EhVHRYLyb44F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebf2c85295ce98718e7df61fc28589f0aeaa73e/pkg/jobs/jobspb/jobs.proto#L474)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

"this this"
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/jobs.proto, line 488 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MQDnWAKTpM3zyX68mnp-r1-488:-MTu_zbG-2epW1aW5DuA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ebf2c85295ce98718e7df61fc28589f0aeaa73e/pkg/jobs/jobspb/jobs.proto#L488)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

"should result in the changefeed restarting"
</blockquote></details>

Done.

**Review by HonoreDB — APPROVED** [review]
> Reviewed 1 of 15 files at r1, 16 of 16 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 3119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MTvXuDV-ynijwkZylq5:-MTvXuDV-ynijwkZylq6:bvkaigv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5698c425e119edc5a590ad9fe5a74e8da6af0da7/pkg/ccl/changefeedccl/changefeed_test.go#L3119)):*
> ```Go
> 		})
> 
> 		// ALTER PRIMARY KEY is totally busted as of writing this.
> ```

Busted no longer?
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MTvaBKu4kctRSpmtFXw:-MTvaBKv6YE5HxdRAgPk:b-izmwa8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5698c425e119edc5a590ad9fe5a74e8da6af0da7/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L277)):*
> ```Go
> 			// index may occur in the same transaction as a change requiring a
> 			// backfill.
> 			if schemafeed.IsPrimaryIndexChange(ev) {
> ```

Is this definitely the right logic? The values haven't changed, true, but the key-value tuples have. Some consumers might end up saving two half-tables this way (data before the schema change and data after), when with a backfill they'd save one half-table and one full table.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58422)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 3119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MTvXuDV-ynijwkZylq5:-MTvuryh441wc4x7PHmZ:b2nh4fg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5698c425e119edc5a590ad9fe5a74e8da6af0da7/pkg/ccl/changefeedccl/changefeed_test.go#L3119)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Busted no longer?
</blockquote></details>

:) 
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MTvaBKu4kctRSpmtFXw:-MTvueYDEy0D2jaE53jI:b-vj0ce3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5698c425e119edc5a590ad9fe5a74e8da6af0da7/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L277)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Is this definitely the right logic? The values haven't changed, true, but the key-value tuples have. Some consumers might end up saving two half-tables this way (data before the schema change and data after), when with a backfill they'd save one half-table and one full table.
</blockquote></details>

I don't think I understand what you're saying regarding half-tables.

**Comment by HonoreDB** [comment]
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MTvaBKu4kctRSpmtFXw:-MTvx-9lDCALBrtAoefJ:b-6xepbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5698c425e119edc5a590ad9fe5a74e8da6af0da7/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L277)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think I understand what you're saying regarding half-tables. 
</blockquote></details>

I'm imagining a consumer that listens to changefeeds with schema_change_policy=backfill on a bunch of tables, and writes values to a giant data lake table with primary key `(table_name, table_version, primary_key)` with buffering and ordering logic so later changes overwrite earlier ones. That way, later, running `Select convert_to_table(*) from megatable GROUP BY (table_name, table_version)` outputs snapshots of the current version of each table, along with snapshots of the state of earlier versions right before the schema change (half-tables). If there's a schema change we don't backfill on, then instead the snapshot of the latest version of the table only includes rows touched after the schema change, and at that point it's awkward to try to merge everything.

**Review by ajwerner — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-:-MU72RWw4KpM2MIp8PRp:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB and @miretskiy)

___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 277 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58422#-MTvaBKu4kctRSpmtFXw:-MU5D_HP3coOToO2-Gdg:b-w8wxmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5698c425e119edc5a590ad9fe5a74e8da6af0da7/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L277)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

I'm imagining a consumer that listens to changefeeds with schema_change_policy=backfill on a bunch of tables, and writes values to a giant data lake table with primary key `(table_name, table_version, primary_key)` with buffering and ordering logic so later changes overwrite earlier ones. That way, later, running `Select convert_to_table(*) from megatable GROUP BY (table_name, table_version)` outputs snapshots of the current version of each table, along with snapshots of the state of earlier versions right before the schema change (half-tables). If there's a schema change we don't backfill on, then instead the snapshot of the latest version of the table only includes rows touched after the schema change, and at that point it's awkward to try to merge everything.
</blockquote></details>

How are they getting access to `table_version`? I remain unclear on how the user does what you're saying -- but we can sort it out later.

**Pull request merged**